# TODO:

1. accuracy CANNOT be 100% (even for SVM??)
1. study performance of models
1. remove bad features (e.g. id)
1. add remaining good features
1. use pandas.get_dummies or scikit OneHotEncoder for one-hot encoding (start with pandas)
1. model and train regression models predicting actual return rate
1. thinks how to continuously train models (every month, or something like that), productionize?



In [1]:
import time

import definitions
import pandas_helper
import scikit_helper
import lendingclub_helper


In [2]:
dataFrames = {}
for key in definitions.dataFiles:
    print("Loading data for " + key)
    dataFrames[key] = pandas_helper.readData(definitions.dataFiles[key])

Loading data for 2007-2011
Loading data for 2012-2013
Loading data for 2014
Loading data for 2015
Loading data for 2016Q1
Loading data for 2016Q2
Loading data for 2016Q3
Loading data for 2016Q4
Loading data for 2017Q1
Loading data for 2017Q2
Loading data for 2017Q3
Loading data for 2017Q4


In [3]:
for key in dataFrames.keys():
    print("Building features for " + key)
    lendingclub_helper.buildFeatures(dataFrames[key])
    featureColumns = lendingclub_helper.getFeatureColumns(dataFrames[key])
    assert(not pandas_helper.columnsHaveNull(dataFrames[key], featureColumns.tolist()))

Building features for 2007-2011
Building features for 2012-2013
Building features for 2014
Building features for 2015
Building features for 2016Q1
Building features for 2016Q2
Building features for 2016Q3
Building features for 2016Q4
Building features for 2017Q1
Building features for 2017Q2
Building features for 2017Q3
Building features for 2017Q4


In [4]:
for key in dataFrames.keys():
    print("Building labels for " + key)
    lendingclub_helper.buildLabels(dataFrames[key])
    print(dataFrames[key].shape)

Building labels for 2007-2011
(42535, 1192)
Building labels for 2012-2013
(188181, 1193)
Building labels for 2014
(235629, 1219)
Building labels for 2015
(421095, 1269)
Building labels for 2016Q1
(133887, 1244)
Building labels for 2016Q2
(97854, 1228)
Building labels for 2016Q3
(99120, 1228)
Building labels for 2016Q4
(103546, 1222)
Building labels for 2017Q1
(96779, 1216)
Building labels for 2017Q2
(105451, 1206)
Building labels for 2017Q3
(122701, 1211)
Building labels for 2017Q4
(118648, 1239)


In [8]:
col = "application_type"
for key in dataFrames.keys():
    print(key)
#     print(dataFrames[key][col].unique().shape)
    print(dataFrames[key][col].unique())


2007-2011
['10+ years' '< 1 year' '1 year' '3 years' '8 years' '9 years' '4 years'
 '5 years' '6 years' '2 years' '7 years' nan]
2012-2013
['4 years' '10+ years' '2 years' '7 years' '1 year' '3 years' '5 years'
 '6 years' '8 years' '< 1 year' '9 years' nan]
2014
['10+ years' '8 years' '< 1 year' '6 years' '2 years' '9 years' '7 years'
 nan '3 years' '5 years' '1 year' '4 years']
2015
['8 years' '2 years' '< 1 year' '10+ years' '1 year' '4 years' '9 years'
 nan '5 years' '6 years' '3 years' '7 years']
2016Q1
['4 years' '3 years' '1 year' nan '10+ years' '6 years' '2 years'
 '< 1 year' '9 years' '8 years' '5 years' '7 years']
2016Q2
['10+ years' '4 years' '5 years' '2 years' nan '6 years' '9 years'
 '3 years' '7 years' '< 1 year' '8 years' '1 year']
2016Q3
['2 years' '10+ years' '3 years' '8 years' '< 1 year' '4 years' '9 years'
 '7 years' '6 years' nan '1 year' '5 years']
2016Q4
['10+ years' '3 years' '5 years' '< 1 year' '2 years' nan '9 years'
 '6 years' '8 years' '4 years' '7 years' 

In [6]:
col = "application_type"
for key in dataFrames.keys():
    print(key)
    cols = [c for c in dataFrames[key].columns if col in c]
    print(dataFrames[key][cols].head())


2007-2011
  application_type  f_application_type_Individual
0       Individual                              1
1       Individual                              1
2       Individual                              1
3       Individual                              1
4       Individual                              1
2012-2013
  application_type  f_application_type_Individual
0       Individual                              1
1       Individual                              1
2       Individual                              1
3       Individual                              1
4       Individual                              1
2014
  application_type  f_application_type_Individual
0       Individual                              1
1       Individual                              1
2       Individual                              1
3       Individual                              1
4       Individual                              1
2015
  application_type  f_application_type_Individual  \
0       Individua

In [11]:
for key in dataFrames.keys():
    print(key)
    print(lendingclub_helper.getFeatureColumns(dataFrames[key]).shape)


2007-2011
(634,)
2012-2013
(623,)
2014
(623,)
2015
(623,)
2016Q1
(634,)
2016Q2
(634,)
2016Q3
(634,)
2016Q4
(623,)
2017Q1
(623,)
2017Q2
(623,)
2017Q3
(623,)
2017Q4
(634,)


In [5]:
for key in dataFrames.keys():
    print("Getting finished loans for " + key)
    labelColumns = lendingclub_helper.getLabelColumns(dataFrames[key])
    dataFrames[key] = lendingclub_helper.getFinishedLoans(dataFrames[key])
    assert(not pandas_helper.columnsHaveNull(dataFrames[key], labelColumns.tolist()))
    print(dataFrames[key].shape)


Getting finished loans for 2007-2011
(42535, 236)
Getting finished loans for 2012-2013
(181326, 236)
Getting finished loans for 2014
(210765, 236)
Getting finished loans for 2015
(245635, 236)
Getting finished loans for 2016Q1
(57531, 236)
Getting finished loans for 2016Q2
(36764, 236)
Getting finished loans for 2016Q3
(34214, 236)
Getting finished loans for 2016Q4
(28809, 236)
Getting finished loans for 2017Q1
(19286, 236)
Getting finished loans for 2017Q2
(13916, 236)
Getting finished loans for 2017Q3
(8986, 236)
Getting finished loans for 2017Q4
(3871, 236)


In [18]:
key1 = "2012-2013"
key2 = "2017Q4"
filename = "../models/svc_{0}.pkl".format(key1)
clf = scikit_helper.loadModel(filename)
df = dataFrames[key2]
features = lendingclub_helper.getFeatureColumns(df)
labels = lendingclub_helper.getLabelColumns(df)
print(getModelAccuracy(clf, df, features, labels))

28
0.9927667269439421


In [11]:
def getModelAccuracy(clf, df, features, labels):
    X = df[features].values
    y = df[labels[0]].values
    y_p = clf.predict(X)
    print(sum(y != y_p))

    total = sum(y == y_p)
    assert(y.shape == y_p.shape)
    return total/len(y)


### Explanation
Rows: what model was trained on

Columns: which result was used to test

Accurary rounded to four decimal places

||2007-2011|2012-2013|2014|2015|2016Q1|2016Q2|2016Q3|2016Q4|2017Q1|2017Q2|2017Q3|2017Q4|
|----|----|----|----|----|----|----|----|----|----|----|----|----|----|
|**2007-2011**|1.0|0.8384|0.8152|0.7474|0.7271|0.7252|0.7464|0.7947|0.8321|0.8594|0.9289|0.9928|
|**2012-2013**|0.8488|1.0|0.8152|0.7474|0.7271|0.7252|0.7464|0.7947|0.8321|0.8594|0.9289|0.9928|
|**2014**|sooka|sooka|sooka|sooka|sooka|sooka|sooka|sooka|sooka|sooka|sooka|sooka|
|**2015**|sooka|sooka|sooka|sooka|sooka|sooka|sooka|sooka|sooka|sooka|sooka|sooka|
|**2016Q1**|sooka|sooka|sooka|sooka|sooka|sooka|sooka|sooka|sooka|sooka|sooka|sooka|
|**2016Q2**|sooka|sooka|sooka|sooka|sooka|sooka|sooka|sooka|sooka|sooka|sooka|sooka|
|**2016Q3**|sooka|sooka|sooka|sooka|sooka|sooka|sooka|sooka|sooka|sooka|sooka|sooka|
|**2016Q4**|sooka|sooka|sooka|sooka|sooka|sooka|sooka|sooka|sooka|sooka|sooka|sooka|
|**2017Q1**|sooka|sooka|sooka|sooka|sooka|sooka|sooka|sooka|sooka|sooka|sooka|sooka|
|**2017Q2**|sooka|sooka|sooka|sooka|sooka|sooka|sooka|sooka|sooka|sooka|sooka|sooka|
|**2017Q3**|sooka|sooka|sooka|sooka|sooka|sooka|sooka|sooka|sooka|sooka|sooka|sooka|
|**2017Q4**|sooka|sooka|sooka|sooka|sooka|sooka|sooka|sooka|sooka|sooka|sooka|sooka|
